File: client-js/client-v2/builders/QueueBuilder.js
/**
 * Queue builder for fluent API
 */

import { generateUUID } from '../../utils/uuid.js'
import { isValidUUID } from '../utils/validation.js'
import { QUEUE_DEFAULTS, CONSUME_DEFAULTS, POP_DEFAULTS } from '../utils/defaults.js'
import * as logger from '../utils/logger.js'

export class QueueBuilder {
  #queen
  #httpClient
  #bufferManager
  #queueName = null
  #partition = 'Default'
  #namespace = null
  #task = null
  #group = null
  #config = {}

  // Consume options
  #concurrency = CONSUME_DEFAULTS.concurrency
  #batch = CONSUME_DEFAULTS.batch
  #limit = CONSUME_DEFAULTS.limit
  #idleMillis = CONSUME_DEFAULTS.idleMillis
  #autoAck = CONSUME_DEFAULTS.autoAck
  #wait = CONSUME_DEFAULTS.wait
  #timeoutMillis = CONSUME_DEFAULTS.timeoutMillis
  #renewLease = CONSUME_DEFAULTS.renewLease
  #renewLeaseIntervalMillis = CONSUME_DEFAULTS.renewLeaseIntervalMillis
  #subscriptionMode = CONSUME_DEFAULTS.subscriptionMode
  #subscriptionFrom = CONSUME_DEFAULTS.subscriptionFrom
  #each = false

  // Buffer options
  #bufferOptions = null

  constructor(queen, httpClient, bufferManager, queueName = null) {
    this.#queen = queen
    this.#httpClient = httpClient
    this.#bufferManager = bufferManager
    this.#queueName = queueName
  }

  // ===========================
  // Queue Configuration Methods
  // ===========================

  namespace(name) {
    this.#namespace = name
    return this
  }

  task(name) {
    this.#task = name
    return this
  }

  config(options) {
    this.#config = { ...QUEUE_DEFAULTS, ...options }
    return this
  }

  create() {
    // Always merge with QUEUE_DEFAULTS to ensure all options are sent
    const fullConfig = Object.keys(this.#config).length > 0 
      ? this.#config 
      : QUEUE_DEFAULTS
    
    const payload = {
      queue: this.#queueName,
      namespace: this.#namespace,
      task: this.#task,
      options: fullConfig
    }

    logger.log('QueueBuilder.create', { queue: this.#queueName, namespace: this.#namespace, task: this.#task })
    return new OperationBuilder(this.#httpClient, 'POST', '/api/v1/configure', payload)
  }

  delete() {
    if (!this.#queueName) {
      throw new Error('Queue name is required for delete operation')
    }

    logger.log('QueueBuilder.delete', { queue: this.#queueName })
    return new OperationBuilder(
      this.#httpClient,
      'DELETE',
      `/api/v1/resources/queues/${encodeURIComponent(this.#queueName)}`,
      null
    )
  }

  // ===========================
  // Push Methods
  // ===========================

  partition(name) {
    this.#partition = name
    return this
  }

  buffer(options) {
    this.#bufferOptions = options
    return this
  }

  push(payload) {
    if (!this.#queueName) {
      throw new Error('Queue name is required for push operation')
    }

    logger.log('QueueBuilder.push', { queue: this.#queueName, partition: this.#partition, count: Array.isArray(payload) ? payload.length : 1, buffered: !!this.#bufferOptions })

    // Format items
    const items = Array.isArray(payload) ? payload : [payload]
    const formattedItems = items.map(item => {
      // Determine the payload - check if property exists, not just truthy
      let payloadValue
      if ('data' in item) {
        payloadValue = item.data
      } else if ('payload' in item) {
        payloadValue = item.payload
      } else {
        payloadValue = item
      }

      const result = {
        queue: this.#queueName,
        partition: this.#partition,
        payload: payloadValue,
        transactionId: item.transactionId || generateUUID()
      }

      // Include traceId if provided and valid UUID
      if (item.traceId && isValidUUID(item.traceId)) {
        result.traceId = item.traceId
      }

      return result
    })

    // Return a PushBuilder for chaining callbacks
    return new PushBuilder(this.#httpClient, this.#bufferManager, this.#queueName, this.#partition, formattedItems, this.#bufferOptions)
  }

  // ===========================
  // Consume Configuration Methods
  // ===========================

  group(name) {
    this.#group = name
    return this
  }

  concurrency(count) {
    this.#concurrency = Math.max(1, count)
    return this
  }

  batch(size) {
    this.#batch = Math.max(1, size)
    return this
  }

  limit(count) {
    this.#limit = count
    return this
  }

  idleMillis(millis) {
    this.#idleMillis = millis
    return this
  }

  autoAck(enabled) {
    this.#autoAck = enabled
    return this
  }

  renewLease(enabled, intervalMillis) {
    this.#renewLease = enabled
    if (intervalMillis) {
      this.#renewLeaseIntervalMillis = intervalMillis
    }
    return this
  }

  subscriptionMode(mode) {
    this.#subscriptionMode = mode
    return this
  }

  subscriptionFrom(from) {
    this.#subscriptionFrom = from
    return this
  }

  each() {
    this.#each = true
    return this
  }

  // ===========================
  // Consume Method
  // ===========================

  consume(handler, options = {}) {
    const consumeOptions = {
      queue: this.#queueName,
      partition: this.#partition !== 'Default' ? this.#partition : null,
      namespace: this.#namespace,
      task: this.#task,
      group: this.#group,
      concurrency: this.#concurrency,
      batch: this.#batch,
      limit: this.#limit,
      idleMillis: this.#idleMillis,
      autoAck: this.#autoAck,
      wait: this.#wait,
      timeoutMillis: this.#timeoutMillis,
      renewLease: this.#renewLease,
      renewLeaseIntervalMillis: this.#renewLeaseIntervalMillis,
      subscriptionMode: this.#subscriptionMode,
      subscriptionFrom: this.#subscriptionFrom,
      each: this.#each,
      signal: options.signal
    }

    return new ConsumeBuilder(this.#httpClient, this.#queen, handler, consumeOptions)
  }

  // ===========================
  // Pop Methods
  // ===========================

  wait(enabled) {
    this.#wait = enabled
    return this
  }

  async pop() {
    logger.log('QueueBuilder.pop', { queue: this.#queueName, partition: this.#partition, namespace: this.#namespace, task: this.#task, batch: this.#batch, wait: this.#wait, group: this.#group })
    
    try {
      const path = this.#buildPopPath()
      
      // For pop(), use POP defaults (not CONSUME defaults)
      // Override autoAck to false unless explicitly set
      const effectiveAutoAck = this.#autoAck !== CONSUME_DEFAULTS.autoAck ? this.#autoAck : POP_DEFAULTS.autoAck
      
      // Build params with correct autoAck for pop
      const params = new URLSearchParams({
        batch: this.#batch.toString(),
        wait: this.#wait.toString(),
        timeout: this.#timeoutMillis.toString()
      })

      if (this.#group) params.append('consumerGroup', this.#group)
      if (this.#namespace) params.append('namespace', this.#namespace)
      if (this.#task) params.append('task', this.#task)
      if (effectiveAutoAck) params.append('autoAck', 'true')
      if (this.#subscriptionMode) params.append('subscriptionMode', this.#subscriptionMode)
      if (this.#subscriptionFrom) params.append('subscriptionFrom', this.#subscriptionFrom)

      const result = await this.#httpClient.get(`${path}?${params}`, this.#timeoutMillis + 5000)

      if (!result || !result.messages) {
        logger.log('QueueBuilder.pop', { status: 'no-messages' })
        return []
      }

      const messages = result.messages.filter(msg => msg != null)
      logger.log('QueueBuilder.pop', { status: 'success', count: messages.length })
      return messages
    } catch (error) {
      // Return empty array on error instead of throwing
      logger.error('QueueBuilder.pop', { error: error.message })
      console.warn('Pop failed:', error.message)
      return []
    }
  }

  #buildPopPath() {
    if (this.#queueName) {
      if (this.#partition && this.#partition !== 'Default') {
        return `/api/v1/pop/queue/${this.#queueName}/partition/${this.#partition}`
      }
      return `/api/v1/pop/queue/${this.#queueName}`
    }

    if (this.#namespace || this.#task) {
      return '/api/v1/pop'
    }

    throw new Error('Must specify queue, namespace, or task for pop operation')
  }

  #buildPopParams() {
    const params = new URLSearchParams({
      batch: this.#batch.toString(),
      wait: this.#wait.toString(),
      timeout: this.#timeoutMillis.toString()  // Server expects 'timeout', not 'timeoutMillis'
    })

    if (this.#group) params.append('consumerGroup', this.#group)
    if (this.#namespace) params.append('namespace', this.#namespace)
    if (this.#task) params.append('task', this.#task)
    if (this.#autoAck) params.append('autoAck', 'true')
    if (this.#subscriptionMode) params.append('subscriptionMode', this.#subscriptionMode)
    if (this.#subscriptionFrom) params.append('subscriptionFrom', this.#subscriptionFrom)

    return params
  }

  // ===========================
  // Buffer Management Methods
  // ===========================

  async flushBuffer() {
    if (!this.#queueName) {
      throw new Error('Queue name is required for buffer flush')
    }
    const queueAddress = `${this.#queueName}/${this.#partition}`
    logger.log('QueueBuilder.flushBuffer', { queueAddress })
    await this.#bufferManager.flushBuffer(queueAddress)
  }

  // ===========================
  // Dead Letter Queue Methods
  // ===========================

  dlq(consumerGroup = null) {
    if (!this.#queueName) {
      throw new Error('Queue name is required for DLQ operations')
    }
    logger.log('QueueBuilder.dlq', { queue: this.#queueName, consumerGroup, partition: this.#partition })
    return new DLQBuilder(this.#httpClient, this.#queueName, consumerGroup, this.#partition)
  }
}

/**
 * Operation builder for create/delete operations with callbacks
 */
class OperationBuilder {
  #httpClient
  #method
  #path
  #body
  #onSuccessCallback = null
  #onErrorCallback = null
  #executed = false

  constructor(httpClient, method, path, body) {
    this.#httpClient = httpClient
    this.#method = method
    this.#path = path
    this.#body = body
  }

  onSuccess(callback) {
    this.#onSuccessCallback = callback
    return this
  }

  onError(callback) {
    this.#onErrorCallback = callback
    return this
  }

  // Auto-execute when awaited
  then(resolve, reject) {
    if (this.#executed) {
      return Promise.resolve().then(resolve, reject)
    }
    this.#executed = true
    return this.#execute().then(resolve, reject)
  }

  async #execute() {
    logger.log('OperationBuilder.execute', { method: this.#method, path: this.#path })
    
    try {
      let result

      if (this.#method === 'GET') {
        result = await this.#httpClient.get(this.#path)
      } else if (this.#method === 'POST') {
        result = await this.#httpClient.post(this.#path, this.#body)
      } else if (this.#method === 'PUT') {
        result = await this.#httpClient.put(this.#path, this.#body)
      } else if (this.#method === 'DELETE') {
        result = await this.#httpClient.delete(this.#path)
      }

      if (result && result.error) {
        const error = new Error(result.error)
        logger.error('OperationBuilder.execute', { method: this.#method, path: this.#path, error: result.error })
        if (this.#onErrorCallback) {
          await this.#onErrorCallback(error)
          return { success: false, error: result.error }
        }
        throw error
      }

      logger.log('OperationBuilder.execute', { method: this.#method, path: this.#path, status: 'success' })
      if (this.#onSuccessCallback) {
        await this.#onSuccessCallback(result)
      }

      return result

    } catch (error) {
      logger.error('OperationBuilder.execute', { method: this.#method, path: this.#path, error: error.message })
      if (this.#onErrorCallback) {
        await this.#onErrorCallback(error)
        return { success: false, error: error.message }
      }
      throw error
    }
  }
}

/**
 * Consume builder for chaining callbacks
 */
class ConsumeBuilder {
  #httpClient
  #queen
  #handler
  #options
  #onSuccessCallback = null
  #onErrorCallback = null
  #executed = false

  constructor(httpClient, queen, handler, options) {
    this.#httpClient = httpClient
    this.#queen = queen
    this.#handler = handler
    this.#options = options
  }

  onSuccess(callback) {
    this.#onSuccessCallback = callback
    return this
  }

  onError(callback) {
    this.#onErrorCallback = callback
    return this
  }

  // Auto-execute when awaited
  then(resolve, reject) {
    if (this.#executed) {
      return Promise.resolve().then(resolve, reject)
    }
    this.#executed = true
    return this.#execute().then(resolve, reject)
  }

  async #execute() {
    // Import ConsumerManager lazily to avoid circular dependency
    const { ConsumerManager } = await import('../consumer/ConsumerManager.js')
    const consumerManager = new ConsumerManager(this.#httpClient, this.#queen)

    // Wrap the handler to include callback logic
    const wrappedHandler = async (msgOrMsgs) => {
      try {
        const result = await this.#handler(msgOrMsgs)
        
        // Call onSuccess if defined
        if (this.#onSuccessCallback) {
          await this.#onSuccessCallback(msgOrMsgs, result)
        }
        
        return result
      } catch (error) {
        // Call onError if defined
        if (this.#onErrorCallback) {
          await this.#onErrorCallback(msgOrMsgs, error)
          // Don't re-throw if callback is defined
          return
        }
        // Re-throw if no error callback
        throw error
      }
    }

    // IMPORTANT: If callbacks are defined, auto-ack must be disabled
    // to prevent double-acking (auto-ack + manual ack in callback)
    const hasCallbacks = this.#onSuccessCallback || this.#onErrorCallback
    const effectiveAutoAck = hasCallbacks ? false : this.#options.autoAck

    const updatedOptions = {
      ...this.#options,
      autoAck: effectiveAutoAck
    }

    return consumerManager.start(wrappedHandler, updatedOptions)
  }
}

/**
 * Push builder for chaining callbacks
 */
class PushBuilder {
  #httpClient
  #bufferManager
  #queueName
  #partition
  #formattedItems
  #bufferOptions
  #onSuccessCallback = null
  #onErrorCallback = null
  #onDuplicateCallback = null
  #executed = false

  constructor(httpClient, bufferManager, queueName, partition, formattedItems, bufferOptions) {
    this.#httpClient = httpClient
    this.#bufferManager = bufferManager
    this.#queueName = queueName
    this.#partition = partition
    this.#formattedItems = formattedItems
    this.#bufferOptions = bufferOptions
  }

  onSuccess(callback) {
    this.#onSuccessCallback = callback
    return this
  }

  onError(callback) {
    this.#onErrorCallback = callback
    return this
  }

  onDuplicate(callback) {
    this.#onDuplicateCallback = callback
    return this
  }

  // Auto-execute when awaited
  then(resolve, reject) {
    if (this.#executed) {
      return Promise.resolve().then(resolve, reject)
    }
    this.#executed = true
    return this.#execute().then(resolve, reject)
  }

  async #execute() {
    logger.log('PushBuilder.execute', { queue: this.#queueName, partition: this.#partition, count: this.#formattedItems.length, buffered: !!this.#bufferOptions })
    
    // Client-side buffering
    if (this.#bufferOptions) {
      for (const item of this.#formattedItems) {
        const queueAddress = `${this.#queueName}/${this.#partition}`
        this.#bufferManager.addMessage(queueAddress, item, this.#bufferOptions)
      }
      const result = { buffered: true, count: this.#formattedItems.length }
      
      logger.log('PushBuilder.execute', { status: 'buffered', count: this.#formattedItems.length })
      
      if (this.#onSuccessCallback) {
        await this.#onSuccessCallback(this.#formattedItems)
      }
      
      return result
    }

    // Immediate push
    try {
      const results = await this.#httpClient.post('/api/v1/push', { items: this.#formattedItems })

      // Server returns an array of results with status for each item
      if (Array.isArray(results)) {
        // Separate results by status
        const successful = []
        const duplicates = []
        const failed = []

        for (let i = 0; i < results.length; i++) {
          const result = results[i]
          const originalItem = this.#formattedItems[i]

          if (result.status === 'duplicate') {
            duplicates.push({ ...originalItem, result })
          } else if (result.status === 'failed') {
            failed.push({ ...originalItem, result, error: result.error })
          } else if (result.status === 'queued') {
            successful.push({ ...originalItem, result })
          }
        }

        // Call appropriate callbacks
        if (duplicates.length > 0 && this.#onDuplicateCallback) {
          await this.#onDuplicateCallback(duplicates, new Error('Duplicate transaction IDs detected'))
        }

        if (failed.length > 0 && this.#onErrorCallback) {
          const error = new Error(failed[0].error || 'Push failed')
          await this.#onErrorCallback(failed, error)
        }

        if (successful.length > 0 && this.#onSuccessCallback) {
          await this.#onSuccessCallback(successful)
        }

        // Only throw if no error callback is defined
        if (failed.length > 0 && !this.#onErrorCallback) {
          logger.error('PushBuilder.execute', { status: 'failed', count: failed.length })
          throw new Error(failed[0].error || 'Push failed')
        }

        logger.log('PushBuilder.execute', { status: 'success', successful: successful.length, duplicates: duplicates.length, failed: failed.length })
        return results
      }

      // Fallback for non-array responses
      if (results && results.error) {
        const error = new Error(results.error)
        if (this.#onErrorCallback) {
          await this.#onErrorCallback(this.#formattedItems, error)
          return results // Don't throw if callback is defined
        }
        throw error
      }

      if (this.#onSuccessCallback) {
        await this.#onSuccessCallback(this.#formattedItems)
      }

      return results

    } catch (error) {
      // Network or HTTP errors
      if (this.#onErrorCallback) {
        await this.#onErrorCallback(this.#formattedItems, error)
        return null // Don't throw if callback is defined
      }
      throw error
    }
  }
}

/**
 * DLQ (Dead Letter Queue) builder for querying failed messages
 */
class DLQBuilder {
  #httpClient
  #queueName
  #consumerGroup
  #partition
  #limit = 100
  #offset = 0
  #from = null
  #to = null

  constructor(httpClient, queueName, consumerGroup, partition) {
    this.#httpClient = httpClient
    this.#queueName = queueName
    this.#consumerGroup = consumerGroup
    this.#partition = partition !== 'Default' ? partition : null
  }

  limit(count) {
    this.#limit = Math.max(1, count)
    return this
  }

  offset(count) {
    this.#offset = Math.max(0, count)
    return this
  }

  from(timestamp) {
    this.#from = timestamp
    return this
  }

  to(timestamp) {
    this.#to = timestamp
    return this
  }

  async get() {
    const params = new URLSearchParams()
    
    params.append('queue', this.#queueName)
    params.append('limit', this.#limit.toString())
    params.append('offset', this.#offset.toString())
    
    if (this.#consumerGroup) {
      params.append('consumerGroup', this.#consumerGroup)
    }
    
    if (this.#partition) {
      params.append('partition', this.#partition)
    }
    
    if (this.#from) {
      params.append('from', this.#from)
    }
    
    if (this.#to) {
      params.append('to', this.#to)
    }

    logger.log('DLQBuilder.get', { queue: this.#queueName, consumerGroup: this.#consumerGroup, partition: this.#partition, limit: this.#limit, offset: this.#offset })

    try {
      const result = await this.#httpClient.get(`/api/v1/dlq?${params}`)
      logger.log('DLQBuilder.get', { status: 'success', total: result?.total || 0, messages: result?.messages?.length || 0 })
      return result || { messages: [], total: 0 }
    } catch (error) {
      logger.error('DLQBuilder.get', { error: error.message })
      console.warn('DLQ query failed:', error.message)
      return { messages: [], total: 0 }
    }
  }
}

File: client-py/queen/builders/queue_builder.py
"""
Queue builder for fluent API
"""

from typing import Any, Callable, Dict, List, Optional, Union
from urllib.parse import urlencode

from ..types import Message
from ..utils import logger
from ..utils.defaults import QUEUE_DEFAULTS, CONSUME_DEFAULTS, POP_DEFAULTS
from ..utils.uuid_gen import generate_uuid
from ..utils.validation import is_valid_uuid
from .consume_builder import ConsumeBuilder
from .dlq_builder import DLQBuilder
from .operation_builder import OperationBuilder
from .push_builder import PushBuilder


class QueueBuilder:
    """Queue builder for fluent API"""

    def __init__(
        self,
        queen: Any,
        http_client: Any,
        buffer_manager: Any,
        queue_name: Optional[str] = None,
    ):
        """
        Initialize queue builder

        Args:
            queen: Queen instance
            http_client: HttpClient instance
            buffer_manager: BufferManager instance
            queue_name: Optional queue name
        """
        self._queen = queen
        self._http_client = http_client
        self._buffer_manager = buffer_manager
        self._queue_name = queue_name
        self._partition = "Default"
        self._namespace: Optional[str] = None
        self._task: Optional[str] = None
        self._group: Optional[str] = None
        self._config: Dict[str, Any] = {}

        # Consume options
        self._concurrency = CONSUME_DEFAULTS["concurrency"]
        self._batch = CONSUME_DEFAULTS["batch"]
        self._limit = CONSUME_DEFAULTS["limit"]
        self._idle_millis = CONSUME_DEFAULTS["idle_millis"]
        self._auto_ack = CONSUME_DEFAULTS["auto_ack"]
        self._wait = CONSUME_DEFAULTS["wait"]
        self._timeout_millis = CONSUME_DEFAULTS["timeout_millis"]
        self._renew_lease = CONSUME_DEFAULTS["renew_lease"]
        self._renew_lease_interval_millis = CONSUME_DEFAULTS["renew_lease_interval_millis"]
        self._subscription_mode = CONSUME_DEFAULTS["subscription_mode"]
        self._subscription_from = CONSUME_DEFAULTS["subscription_from"]
        self._each = False

        # Buffer options
        self._buffer_options: Optional[Dict[str, Any]] = None

    # ===========================
    # Queue Configuration Methods
    # ===========================

    def namespace(self, name: str) -> "QueueBuilder":
        """Set namespace"""
        self._namespace = name
        return self

    def task(self, name: str) -> "QueueBuilder":
        """Set task"""
        self._task = name
        return self

    def config(self, options: Dict[str, Any]) -> "QueueBuilder":
        """Set queue configuration"""
        self._config = {**QUEUE_DEFAULTS, **options}
        return self

    def create(self) -> OperationBuilder:
        """Create queue"""
        # Always merge with QUEUE_DEFAULTS to ensure all options are sent
        full_config = self._config if self._config else QUEUE_DEFAULTS

        payload = {
            "queue": self._queue_name,
            "namespace": self._namespace,
            "task": self._task,
            "options": full_config,
        }

        logger.log(
            "QueueBuilder.create",
            {"queue": self._queue_name, "namespace": self._namespace, "task": self._task},
        )
        return OperationBuilder(self._http_client, "POST", "/api/v1/configure", payload)

    def delete(self) -> OperationBuilder:
        """Delete queue"""
        if not self._queue_name:
            raise ValueError("Queue name is required for delete operation")

        logger.log("QueueBuilder.delete", {"queue": self._queue_name})
        return OperationBuilder(
            self._http_client,
            "DELETE",
            f"/api/v1/resources/queues/{self._queue_name}",
            None,
        )

    # ===========================
    # Push Methods
    # ===========================

    def partition(self, name: str) -> "QueueBuilder":
        """Set partition"""
        self._partition = name
        return self

    def buffer(self, options: Dict[str, Any]) -> "QueueBuilder":
        """Enable client-side buffering"""
        self._buffer_options = options
        return self

    def push(self, payload: Union[Dict[str, Any], List[Dict[str, Any]]]) -> PushBuilder:
        """Push messages"""
        if not self._queue_name:
            raise ValueError("Queue name is required for push operation")

        logger.log(
            "QueueBuilder.push",
            {
                "queue": self._queue_name,
                "partition": self._partition,
                "count": len(payload) if isinstance(payload, list) else 1,
                "buffered": self._buffer_options is not None,
            },
        )

        # Format items
        items = payload if isinstance(payload, list) else [payload]
        formatted_items = []

        for item in items:
            # Determine the payload - check if property exists, not just truthy
            if "data" in item:
                payload_value = item["data"]
            elif "payload" in item:
                payload_value = item["payload"]
            else:
                payload_value = item

            result = {
                "queue": self._queue_name,
                "partition": self._partition,
                "payload": payload_value,
                "transactionId": item.get("transactionId") or generate_uuid(),
            }

            # Include traceId if provided and valid UUID
            trace_id = item.get("traceId")
            if trace_id and is_valid_uuid(trace_id):
                result["traceId"] = trace_id

            formatted_items.append(result)

        # Return a PushBuilder for chaining callbacks
        return PushBuilder(
            self._http_client,
            self._buffer_manager,
            self._queue_name,
            self._partition,
            formatted_items,
            self._buffer_options,
        )

    # ===========================
    # Consume Configuration Methods
    # ===========================

    def group(self, name: str) -> "QueueBuilder":
        """Set consumer group"""
        self._group = name
        return self

    def concurrency(self, count: int) -> "QueueBuilder":
        """Set concurrency"""
        self._concurrency = max(1, count)
        return self

    def batch(self, size: int) -> "QueueBuilder":
        """Set batch size"""
        self._batch = max(1, size)
        return self

    def limit(self, count: int) -> "QueueBuilder":
        """Set message limit"""
        self._limit = count
        return self

    def idle_millis(self, millis: int) -> "QueueBuilder":
        """Set idle timeout"""
        self._idle_millis = millis
        return self

    def auto_ack(self, enabled: bool) -> "QueueBuilder":
        """Set auto-ack"""
        self._auto_ack = enabled
        return self

    def renew_lease(self, enabled: bool, interval_millis: Optional[int] = None) -> "QueueBuilder":
        """Enable lease renewal"""
        self._renew_lease = enabled
        if interval_millis:
            self._renew_lease_interval_millis = interval_millis
        return self

    def subscription_mode(self, mode: str) -> "QueueBuilder":
        """Set subscription mode"""
        self._subscription_mode = mode
        return self

    def subscription_from(self, from_: str) -> "QueueBuilder":
        """Set subscription start point"""
        self._subscription_from = from_
        return self

    def each(self) -> "QueueBuilder":
        """Process messages one at a time"""
        self._each = True
        return self

    # ===========================
    # Consume Method
    # ===========================

    def consume(
        self,
        handler: Callable[[Union[Message, List[Message]]], Any],
        *,
        signal: Optional[Any] = None,  # asyncio.Event
    ) -> ConsumeBuilder:
        """
        Start consuming messages

        Args:
            handler: Message handler
            signal: Optional abort signal (asyncio.Event)

        Returns:
            ConsumeBuilder for chaining
        """
        consume_options = {
            "queue": self._queue_name,
            "partition": self._partition if self._partition != "Default" else None,
            "namespace": self._namespace,
            "task": self._task,
            "group": self._group,
            "concurrency": self._concurrency,
            "batch": self._batch,
            "limit": self._limit,
            "idle_millis": self._idle_millis,
            "auto_ack": self._auto_ack,
            "wait": self._wait,
            "timeout_millis": self._timeout_millis,
            "renew_lease": self._renew_lease,
            "renew_lease_interval_millis": self._renew_lease_interval_millis,
            "subscription_mode": self._subscription_mode,
            "subscription_from": self._subscription_from,
            "each": self._each,
            "signal": signal,
        }

        return ConsumeBuilder(self._http_client, self._queen, handler, consume_options)

    # ===========================
    # Pop Methods
    # ===========================

    def wait(self, enabled: bool) -> "QueueBuilder":
        """Enable/disable long polling"""
        self._wait = enabled
        return self

    async def pop(self) -> List[Dict[str, Any]]:
        """Pop messages"""
        logger.log(
            "QueueBuilder.pop",
            {
                "queue": self._queue_name,
                "partition": self._partition,
                "namespace": self._namespace,
                "task": self._task,
                "batch": self._batch,
                "wait": self._wait,
                "group": self._group,
            },
        )

        try:
            path = self._build_pop_path()

            # For pop(), use POP defaults (not CONSUME defaults)
            # Override autoAck to false unless explicitly set
            effective_auto_ack = (
                self._auto_ack
                if self._auto_ack != CONSUME_DEFAULTS["auto_ack"]
                else POP_DEFAULTS["auto_ack"]
            )

            # Build params with correct autoAck for pop
            params: Dict[str, str] = {
                "batch": str(self._batch),
                "wait": str(self._wait).lower(),
                "timeout": str(self._timeout_millis),
            }

            if self._group:
                params["consumerGroup"] = self._group
            if self._namespace:
                params["namespace"] = self._namespace
            if self._task:
                params["task"] = self._task
            if effective_auto_ack:
                params["autoAck"] = "true"
            if self._subscription_mode:
                params["subscriptionMode"] = self._subscription_mode
            if self._subscription_from:
                params["subscriptionFrom"] = self._subscription_from

            # Generate affinity key for consistent routing to same backend
            affinity_key = self._get_affinity_key()

            query_string = urlencode(params)
            result = await self._http_client.get(
                f"{path}?{query_string}", self._timeout_millis + 5000, affinity_key
            )

            if not result or not result.get("messages"):
                logger.log("QueueBuilder.pop", {"status": "no-messages"})
                return []

            messages = [msg for msg in result["messages"] if msg is not None]
            logger.log("QueueBuilder.pop", {"status": "success", "count": len(messages)})
            return messages
        except Exception as error:
            # Return empty array on error instead of throwing
            logger.error("QueueBuilder.pop", {"error": str(error)})
            print(f"Pop failed: {error}")
            return []

    def _build_pop_path(self) -> str:
        """Build pop path"""
        if self._queue_name:
            if self._partition and self._partition != "Default":
                return f"/api/v1/pop/queue/{self._queue_name}/partition/{self._partition}"
            return f"/api/v1/pop/queue/{self._queue_name}"

        if self._namespace or self._task:
            return "/api/v1/pop"

        raise ValueError("Must specify queue, namespace, or task for pop operation")

    # ===========================
    # Affinity Key Generation
    # ===========================

    def _get_affinity_key(self) -> Optional[str]:
        """
        Generate affinity key for consistent routing
        Matches server's PollIntention::grouping_key() format
        Format: queue:partition:consumerGroup or namespace:task:consumerGroup
        """
        if self._queue_name:
            # Queue-based routing: queue:partition:consumerGroup
            partition = self._partition or "*"
            group = self._group or "__QUEUE_MODE__"
            return f"{self._queue_name}:{partition}:{group}"
        elif self._namespace or self._task:
            # Namespace/task-based routing: namespace:task:consumerGroup
            namespace = self._namespace or "*"
            task = self._task or "*"
            group = self._group or "__QUEUE_MODE__"
            return f"{namespace}:{task}:{group}"
        return None

    # ===========================
    # Buffer Management Methods
    # ===========================

    async def flush_buffer(self) -> None:
        """Flush buffer for this queue"""
        if not self._queue_name:
            raise ValueError("Queue name is required for buffer flush")
        queue_address = f"{self._queue_name}/{self._partition}"
        logger.log("QueueBuilder.flushBuffer", {"queue_address": queue_address})
        await self._buffer_manager.flush_buffer(queue_address)

    # ===========================
    # Dead Letter Queue Methods
    # ===========================

    def dlq(self, consumer_group: Optional[str] = None) -> DLQBuilder:
        """Query dead letter queue"""
        if not self._queue_name:
            raise ValueError("Queue name is required for DLQ operations")
        logger.log(
            "QueueBuilder.dlq",
            {"queue": self._queue_name, "consumer_group": consumer_group, "partition": self._partition},
        )
        return DLQBuilder(self._http_client, self._queue_name, consumer_group, self._partition)

